openedx/core/openapi.py

```diff
@@ -44,22 +49,108 @@ def dedent(text):
     return textwrap.dedent(text)
 
 
-def swagger_auto_schema(**kwargs):
+def schema(*args, **kwargs):
     """
     Decorator for documenting an OpenAPI endpoint.
 
-    Identical to `drf_yasg.utils.swagger_auto_schema`__ except that
-    description fields will be dedented properly.  All description fields
-    should be in Markdown.
+    Identical to `drf_yasg.utils.swagger_auto_schema`__ except that:
+    * If in kwargs, operation_description and operation_summary will be
+      dedented properly.
+    * Otherwise, those fields will be taken from the docstring.
+    * Any values in `view_fun._openapi_decorator_data` will be used as
+      overrides on kwargs.
 
     __ https://drf-yasg.readthedocs.io/en/stable/drf_yasg.html#drf_yasg.utils.swagger_auto_schema
 
     """
-    if 'operation_description' in kwargs:
-        kwargs['operation_description'] = dedent(kwargs['operation_description'])
-    for param in kwargs.get('manual_parameters', ()):
-        param.description = dedent(param.description)
-    return drf_swagger_auto_schema(**kwargs)
+    if args and callable(args[0]):
+        # decorator may be used with no argument
+        return schema(*args[1:], **kwargs)(args[0])
+
+    def decorator(view_func):
+        """ Returns the decorated function. """
+        # pylint: disable=protected-access
+        if view_func.__doc__ is not None:
+            doc_lines = view_func.__doc__.strip().split("\n")
+            if 'operation_summary' not in kwargs and doc_lines:
+                kwargs['operation_summary'] = doc_lines[0].strip()
+            if 'operation_description' not in kwargs and len(doc_lines) > 1:
+                kwargs['operation_description'] = "\n".join(doc_lines[1:])
+        if 'operation_description' in kwargs:
+            kwargs['operation_description'] = dedent(kwargs['operation_description'])
+        try:
+            decorator_data = view_func._openapi_decorator_data
+        except AttributeError:
+            pass
+        else:
+            kwargs.update(decorator_data)
```

Comment:
Isn't this going to fail whenever an AttributeError exception is raised? AFAIU the `decorator_data` will be undefined.

Reply:
The `else` clause is only executed if there is no exception caught.

Reply:
The `else` clause is only executed if there is no exception caught.

Reply:
The `else` clause is only executed if there is no exception caught.

Reply:
Right, my bad. I suggest we get rid of the "else" with:

        try:
            kwargs.update(view_func._openapi_decorator_data)
        except AttributeError:
            pass

Reply:
I like the `else` pattern because it makes the `AttributeError`-catching scope as minimal as possible.  With `kwargs.update` in the try, bugs like `kwargs` accidentally being set to a string or list would pass silently.

If we were to stick with the stacked decorators and use this code, I could put a comment to explain the `else` clause, as it's not something you see used a lot in Python.